core/shared/src/main/scala/zio/internal/FiberContext.scala

```diff
@@ -413,13 +413,13 @@ private[zio] final class FiberContext[E, A](
                   val zio = curZio.asInstanceOf[ZIO.EffectAsync[Any, E, Any]]
 
                   // Enter suspended state:
-                  curZio = if (enterAsync()) {
+                  curZio = if (enterAsync(curZio)) {
```

Comment:
You have to be careful to avoid the capture of `curZio` inside a lambda, as this will create an `ObjectRef` that forces access to the variable to go through the heap. In these cases you can use `zio` to avoid any possibility of capturing the wrong value.